### ISSUE TITLE
package.json "exports" not respected when .ts file exists for subpath

### ISSUE THREAD

**Issue by aleclarson** [microsoft/TypeScript#51172]
# Bug Report

Reproduction:
https://github.com/aleclarson/repro/tree/ts-exports

### 🔎 Search Terms

- package.json exports
- moduleResolution node16 nodenext
- pnpm workspace
- nested package

### 🕗 Version & Regression Information

Tested with `typescript@4.8.4` and `4.9.0-dev.20221013`

### 💻 Code

https://github.com/aleclarson/repro/tree/ts-exports

### 🙁 Actual behavior

The `@pkg/foo` specifier resolves to `src/pkg/foo.ts`

### 🙂 Expected behavior

The `@pkg/foo` specifier resolves to `src/pkg/dist/foo.d.ts`

**Comment by weswigham** [comment]
This is intentional? Since the path in the export map is the output path for a file in your program, it resolved to the input file instead.

**Comment by aleclarson** [comment]
> the path in the export map is the output path for a file in your program

I don't think it is? The `src/pkg` folder is excluded by `tsconfig.json` and a bare import is being used, so it should behave as if the package is a regular dependency (ie: `exports` field should be used instead of `.ts` file).

**Comment by weswigham** [comment]
Yeah, `exclude` doesn't quite work the way your usage implies. `exclude` removes `src/foo` from the root files list, but it's still loaded into your program via it's use in `main`, so it's files still have a reverse mapping of output -> input. Anyways, why do you care that it resolves to the output file in our resolver, anyway? It doesn't affect filepaths or anything, and if you wanted to be using the declaration files for language service speed, you should be using project references instead.

**Comment by aleclarson** [comment]
> but it's still loaded into your program via its use in `main`, so its files still have a reverse mapping of output -> input

That seems counter-intuitive. I'm using it like a dependency, but it's treated as a relative import simply because it exists in the `src` folder?

> why do you care that it resolves to the output file in our resolver? you should be using project references instead

I tried that first, but it didn't work for my use case. The `src/pkg` folder relies on `dom` lib, but also imports utility functions (from `src/utils`) which are imported by modules both in `src` and `src/pkg` (the two folders containing `tsconfig.json`). Therefore, I have to make `src/utils` an external package or else I get the error about importing outside the `rootDir` when a `src/pkg` module imports a `src/utils` module.

**Comment by weswigham** [comment]
We also quietly choose not to emit source files found under `node_modules` because some packages shipped `.ts` sources for awhile, which we prefer to load over `.d.ts` files, which is why it may look like it's "not included in the build" (because only the non-symlinked build actually triggers emit).

**Comment by weswigham** [comment]
> Therefore, I have to make src/utils an external package or else I get the error about importing outside the rootDir when a src/pkg module imports a src/utils module.

You can just set the `rootDir` option to `../`, afaik. `composite` just defaults it to wherever the tsconfig is.

**Comment by aleclarson** [comment]
> You can just set the `rootDir` option to `../`, afaik

No effect, it seems.
https://github.com/aleclarson/repro/tree/ts-exports-fix-1

**Comment by aleclarson** [comment]
> > but it's still loaded into your program via its use in `main`, so its files still have a reverse mapping of output -> input
> 
> That seems counter-intuitive. I'm using it like a dependency, but it's treated as a relative import simply because it exists in the `src` folder?

Is there an explanation for this?

**Comment by weswigham** [comment]
It's simple enough, really - all `.ts` files found during resolution are considered candidates for emit, and are emitted, except for the `node_modules` member exception.

**Comment by aleclarson** [comment]
> It's simple enough, really - all `.ts` files found during resolution are considered candidates for emit, and are emitted, except for the `node_modules` member exception.

I'm confused how it's finding `src/pkg/foo.ts` in the first place. I'm importing `@pkg/foo`, which should resolve to `src/pkg/dist/foo.d.ts` due to exports field resolution. What am I missing?
